astropy/io/fits/src/compressionmodule.c

```diff
@@ -1016,6 +1016,96 @@ PyObject* compression_compress_hdu(PyObject* self, PyObject* args)
 }
 
 
+// This is a cleaned-up version of the cfitio 3.41 function
+// 'fits_get_img_equivtype' (AKA 'ffgiet') with a slightly different interface.
+// It's necessary to copy this code because the original code reads the header,
+// but our munged fitsfile doesn't have any header to read.
+static void get_equiv_datatype(
+    int bitpix,  // BITPIX for image
+    double bscale,  // BSCALE for image
+    double bzero,  // BZERO for image
+    int *imgtype,  // cfitsio image type to use
+    int *numpytype  // numpy image type to use
+    )
+{
+    double min_val;
+    double max_val;
+    long lngzero;
+    long lngscale;
+
+    bitpix_to_datatypes(bitpix, imgtype, numpytype);
+
+    if (bscale == 1.0 && bzero == 0.0) {
+        return;  /* no scaling */
+    }
+
+    switch (*imgtype)
+    {
+      case BYTE_IMG:   /* 8-bit image */
+        min_val = 0.;
+        max_val = 255.0;
+        break;
+      case SHORT_IMG:
+        min_val = -32768.0;
+        max_val =  32767.0;
+        break;
+      case LONG_IMG:
+        min_val = -2147483648.0;
+        max_val =  2147483647.0;
+        break;
+      default:  /* don't have to deal with other data types */
+        return;
+    }
+
+    if (bscale >= 0.) {
+        min_val = bzero + bscale * min_val;
+        max_val = bzero + bscale * max_val;
+    } else {
+        max_val = bzero + bscale * min_val;
+        min_val = bzero + bscale * max_val;
+    }
+    if (bzero < 2147483648.) {  /* don't exceed range of 32-bit integer */
+       lngzero = (long) bzero;
+    }
+    lngscale = (long) bscale;
+
+    if ((bzero != 2147483648.) && /* special value that exceeds integer range */
+       (lngzero != bzero || lngscale != bscale)) { /* not integers? */
+       /* floating point scaled values; just decide on required precision */
+        if (*imgtype == BYTE_IMG || *imgtype == SHORT_IMG) {
+            *imgtype = FLOAT_IMG;
+            *numpytype = NPY_FLOAT;
+        } else {
+            *imgtype = DOUBLE_IMG;
+            *numpytype = NPY_DOUBLE;
+        }
+    /*
+       In all the remaining cases, BSCALE and BZERO are integers,
+       and not equal to 1 and 0, respectively.
+    */
+    } else if ((min_val == -128.) && (max_val == 127.)) {
+        *imgtype = SBYTE_IMG;
+        *numpytype = NPY_INT8;
+    } else if ((min_val >= -32768.0) && (max_val <= 32767.0)) {
+        *imgtype = SHORT_IMG;
+        *numpytype = NPY_INT16;
+    } else if ((min_val >= 0.0) && (max_val <= 65535.0)) {
+        *imgtype = USHORT_IMG;
```

Comment:
I wonder if we should really support unsigned integers here. It seems like we haven't previously.